### PR TITLE
luna: add block iterator

### PR DESCRIPTION
**Pull request by PatrickNicholas**
ref #221

### CODE REVIEW

**Review by huachaohuang — APPROVED** [review]
The iterator interfaces look fine. Some details can be improved. But let's move forward to see how things work together first.

**Comment by huachaohuang** [comment]
@PatrickNicholas You can continue with `TableIter`/`TableReader`. You are going to need #231 #232 . So you can review that first.

**Comment by w41ter-l** [comment]
@huachaohuang Ok, I can try `TableReader` first.

File: src/engine/luna/src/table/block_iter.rs [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+fn read_value(data: &[u8], offset: usize) -> &[u8] {
+    &data[value_range(data, offset)]
+}
+
+fn next_entry_offset(data: &[u8], offset: usize) -> usize {
+    value_range(data, offset).end
+}
+
+pub struct BlockIter {
+    data: Arc<Vec<u8>>,
```
Comment (skyzh) [inline src/engine/luna/src/table/block_iter.rs:65]: Using Arc<[u8]> can reduce a layer of indirection.

File: src/engine/luna/src/table/block_iter.rs [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+    let offset = data.len() - core::mem::size_of::<u32>();
+    (&data[offset..]).get_u32()
+}
+
+fn read_restart(restarts: &[u8], index: usize) -> u32 {
+    let offset = index * core::mem::size_of::<u32>();
+    let end = offset + core::mem::size_of::<u32>();
+
+    debug_assert!(restarts.len() >= end);
+    (&restarts[offset..end]).get_u32()
```
Comment (skyzh) [inline src/engine/luna/src/table/block_iter.rs:31]: The range index operator will check for overflow. No need to debug_assert again.

**Comment by huachaohuang** [comment]
@skyzh Thanks for the comments. We will improve the code quality when the related parts work well together.

**Pull request merged**